### PR TITLE
Use unix sockets instead of pipes for posix daemon.

### PR DESCRIPTION
**Pull request by m-ou-se**
**Summary**

This replaces the way clients communicate with the daemon on Posix.
Instead of every client writing to the same pipe, and creating a separate pipe back for each client, this uses proper (unix socket) *connections* to handle multiple clients, which simplifies a few things.

**Detail**

Unlike pipes, unix sockets provide bi-directional
communication with each connected client.

- No need to generate a unique uuid per client anymore.

- The client doesn't have to create its own pipe anymore.

- Since there is no risk of multiple client's writes getting mixed up,
  messages don't need to fit in a single write anymore, removing the
  limit on command length.

- Since the server can detect a connection closing, the client no longer
  needs to handle signals. When the client is killed, the connection is
  automatically closed, which will cause the server to kill the related
  px4 thread.
  Since this does not rely on handling signals and the client sending an
  additional message, this is much more reliable.

- Client is no longer a singleton.

- The protocol is simplified. Standard output is directly written to the
  socket back to the client, without wrapping it in any protocol
  message.

- Because of the simple protocol, one could now even use netcat to run a
  px4 command:

      $ echo hello | netcat -UN /tmp/px4-sock-0

Also removes a few race conditions.

### CODE REVIEW

File: platforms/posix/src/px4_daemon/client.cpp [hunk 1]
```diff
@@ -116,70 +92,43 @@ Client::process_args(const int argc, const char **argv)
...
+	const char *buf = cmd_buf.data();
 
-	// The size is +1 because we want to include the null termination.
-	packet.header.payload_length = cmd_buf.size() + 1 + sizeof(packet.payload.execute_msg.is_atty);
+	while (n > 0) {
+		int n_sent = write(_fd, buf, n);
 
-	_client_send_pipe_fd = open(get_client_send_pipe_path(_instance_id).c_str(), O_WRONLY);
+		if (n_sent < 0) {
+			PX4_ERR("write() failed: ", strerror(errno));
```
[the remark this answers is not in the document]
Comment (m-ou-se) [reply platforms/posix/src/px4_daemon/client.cpp]: Good catch. Though, the compiler should probably have caught it automatically.

Working on https://github.com/PX4/Firmware/pull/10773 now to fix this problem in the future.

(Already found more than 30 other places in the currently checked in code where format strings are wrong.)

**Comment by m-ou-se** [comment]
The performance is hard to measure, as the time it takes to start and stop the client process is much more significant than the time it takes to connect and communicate. Just running one process with the `Client` class in a loop should show a difference, but the result is not realistic, as the `Client` is only ever used through a separate client process.

The few simple tests I ran using the client programs as they are, show no significant difference in performance. CPU usage is about the same, number of requests handled per second is about the same, etc.

I ran a stress test for several minutes, with multiple terminals running commands concurrently. There was only a few percent cpu usage by the daemon, and memory usage was low and stable.

File: platforms/posix/src/px4_daemon/client.cpp [hunk 2]
```diff
@@ -188,159 +137,53 @@ Client::_send_cmds(const int argc, const char **argv)
...
-			// 0 means the pipe has been closed by all clients.
-			exit_arg = 0;
-			exit_loop = true;
+		} else {
+			n_read += n_buffer_used;
+
+			if (n_read >= 2 && buffer[n_read - 2] == 0) {
+				// If the buffer ends in {0, retval}, keep it.
+				fwrite(buffer, n_read - 2, 1, stdout);
+				buffer[0] = 0;
```
[the remark this answers is not in the document]
Comment (m-ou-se) [reply platforms/posix/src/px4_daemon/client.cpp]: The return should not have been there. Only the `n_read == 0` case above should return. Fixed.

File: platforms/posix/src/px4_daemon/client.cpp [hunk 3]
```diff
@@ -188,159 +137,53 @@ Client::_send_cmds(const int argc, const char **argv)
...
+				buffer[1] = buffer[n_read - 1];
+				n_buffer_used = 2;
+				return buffer[n_read - 1];
```
[the remark this answers is not in the document]
Comment (m-ou-se) [reply platforms/posix/src/px4_daemon/client.cpp]: Not a problem now the return statement is gone, but I moved it to the destructor anyway.

File: platforms/posix/src/px4_daemon/server.h [hunk 4]
```diff
@@ -114,28 +103,19 @@ class Server
...
+	static void *_handle_client(void *arg);
+	static void _cleanup(int fd);
 
 	pthread_t _server_main_pthread;
 
-	std::map<pthread_t, int> _pthread_to_pipe_fd;
-	std::map<uint64_t, pthread_t> _client_uuid_to_pthread;
-	pthread_mutex_t _mutex; ///< protects access to _pthread_to_pipe_fd and _client_uuid_to_pthread
+	std::map<int, pthread_t> _fd_to_thread;
+	pthread_mutex_t _mutex;
```
[the remark this answers is not in the document]
Comment (m-ou-se) [reply platforms/posix/src/px4_daemon/server.h]: Done.

**Comment by m-ou-se** [comment]
@bkueng I can't reproduce the problem of `px4-uorb top` exiting here. If `uorb top` prints any nul-bytes, the return statement that's now removed could have cause this. Otherwise, we'll need to investigate more.

Can you trigger it without using jmavsim? That'd make it easier for me to reproduce.

**Comment by julianoes** [comment]
Looks like this wasn't tested on macOS :sob: . #10839

File: platforms/posix/src/px4_daemon/server.cpp [hunk 5]
```diff
@@ -35,29 +35,27 @@
...
 #include <sys/types.h>
+#include <sys/un.h>
+#include <vector>
 
 #include <px4_log.h>
 
 #include "pxh.h"
 #include "server.h"
 
-// In Cygwin opening and closing the same named pipe multiple times within one process doesn't work POSIX compliant.
```
Comment (MaEtUgR) [inline platforms/posix/src/px4_daemon/server.cpp]: @m-ou-se **TOTALLY AWESOME!** 👍 👍  Thanks a lot for switching to sockets. I planned myself to attempt an implementation with this approach and did some research for that. Sockets have in my eyes only advantages for this use case, your commit message https://github.com/PX4/Firmware/commit/9594ebf72efdfff1a88aaf2206a90836ee9b9130 says it all. Also my main motivation would have been to get rid of this Cygwin hack (came in https://github.com/PX4/Firmware/pull/10371) to make it work under Windows. Thanks again, it works great on Windows now without any hack.

Comment (julianoes) [reply platforms/posix/src/px4_daemon/server.cpp]: Oh that's good news @MaEtUgR.

**Pull request merged**